### PR TITLE
Upgrade to Cairo `2.5.3` and other refactors

### PR DESCRIPTION
**Pull request by delaaxe**
## Pull Request type

Please check the type of change your PR introduces:

- [X] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by LucasLvy** [comment]
put #272 and #270 in the same PR as this one

**Comment by delaaxe** [comment]
Sure

**Comment by delaaxe** [comment]
@LucasLvy done

File: Scarb.toml [hunk 1]
```diff
@@ -19,7 +19,7 @@ description = "Community maintained Cairo and Starknet libraries"
 homepage = "https://github.com/keep-starknet-strange/alexandria/"
 
 [workspace.dependencies]
-starknet = ">=2.4.0"
+starknet = ">=2.5.0"
```
Comment (LucasLvy) [inline Scarb.toml:22]: Shouldn't it be this ?
```suggestion
starknet = ">=2.5.3"
```

File: src/ascii/src/integer.cairo [hunk 2]
```diff
@@ -71,11 +68,9 @@ impl SmallIntegerToAsciiTraitImpl<
...
         let mut inverse_ascii_arr = self.to_inverse_ascii_array().span();
         let mut ascii: felt252 = 0;
-        loop {
-            match inverse_ascii_arr.pop_back() {
-                Option::Some(val) => { ascii = ascii * 256 + *val; },
-                Option::None(_) => { break; },
-            };
+        while !inverse_ascii_arr.is_empty() {
+            let val = *inverse_ascii_arr.pop_back().unwrap();
+            ascii = ascii * 256 + val;
```
Comment (LucasLvy) [inline src/ascii/src/integer.cairo:73]: revert this change i think the previous version is more optimised

File: src/bytes/src/bytes copy.cairo [hunk 3]
```diff
@@ -0,0 +1,507 @@
+use alexandria_bytes::utils::{
```
Comment (LucasLvy) [inline src/bytes/src/bytes copy.cairo:1]: what is this file?

File: src/bytes/src/utils.cairo [hunk 4]
```diff
@@ -6,17 +6,13 @@ use keccak::{u128_to_u64, u128_split as u128_split_to_u64, cairo_keccak};
...
-                keccak_add_uint128_be(ref keccak_input, *v, value_size);
-                size -= value_size;
-            },
-            Option::None => { break; },
-        };
+    while !input.is_empty() {
+        let v = *input.pop_front().unwrap();
+        let value_size = uint_min(size, 16);
+        keccak_add_uint128_be(ref keccak_input, v, value_size);
+        size -= value_size;
```
Comment (LucasLvy) [inline src/bytes/src/utils.cairo:15]: revert that

File: src/linalg/src/kron.cairo [hunk 5]
```diff
@@ -22,18 +22,12 @@ fn kron<T, +Mul<T>, +AddEq<T>, +Zeroable<T>, +Copy<T>, +Drop<T>,>(
...
-                    };
-                };
-            },
-            Option::None => { break; },
+    while !xs.is_empty() {
+        let x = *xs.pop_front().unwrap();
+        let mut ys_copy = ys;
+        while !ys_copy.is_empty() {
+            let y = *ys_copy.pop_front().unwrap();
+            array.append(x * y);
```
Comment (LucasLvy) [inline src/linalg/src/kron.cairo:30]: revert

File: src/linalg/src/norm.cairo [hunk 6]
```diff
@@ -13,19 +13,15 @@ fn norm<T, +Into<T, u128>, +Zeroable<T>, +Copy<T>>(
...
-        };
+    while !xs.is_empty() {
+        let x = xs.pop_front().unwrap();
+        if ord == 0 {
+            if (*x).is_non_zero() {
+                norm += 1;
+            }
+        } else {
+            norm += pow((*x).into(), ord);
+        }
```
Comment (LucasLvy) [inline src/linalg/src/norm.cairo:24]: revert

File: src/math/src/gcd_of_n_numbers.cairo [hunk 7]
```diff
@@ -11,12 +11,11 @@ fn gcd(mut n: Span<u128>) -> u128 {
...
-        match n.pop_front() {
-            Option::Some(b) => { a = gcd_two_numbers(a, *b); },
-            Option::None => { break a; },
-        };
-    }
+    while !n.is_empty() {
+        let b = *n.pop_front().unwrap();
+        a = gcd_two_numbers(a, b);
+    };
+    a
```
Comment (LucasLvy) [inline src/math/src/gcd_of_n_numbers.cairo:18]: revert

File: src/math/src/lcm_of_n_numbers.cairo [hunk 8]
```diff
@@ -22,13 +19,10 @@ fn lcm<T, +Into<T, u128>, +Into<u128, T>, +Mul<T>, +Div<T>, +Copy<T>, +Drop<T>>(
...
-            },
-            Option::None => { break Result::Ok(a); },
-        };
-    }
+    while !n.is_empty() {
+        let b = *n.pop_front().unwrap();
+        let gcd: T = gcd_two_numbers(a.into(), b.into()).into();
+        a = (a * b) / gcd;
+    };
+    Result::Ok(a)
```
Comment (LucasLvy) [inline src/math/src/lcm_of_n_numbers.cairo:27]: revert

File: src/numeric/src/cumprod.cairo [hunk 9]
```diff
@@ -10,18 +10,13 @@ fn cumprod<T, +Mul<T>, +Copy<T>, +Drop<T>,>(mut sequence: Span<T>) -> Array<T> {
...
-                prev_value = prod;
-            },
-            Option::None => { break; },
-        };
+    let mut array = array![prev_value];
+    while !sequence.is_empty() {
+        let current_value = *sequence.pop_front().unwrap();
+        let product = current_value * prev_value;
+        array.append(product);
+        prev_value = product;
```
Comment (LucasLvy) [inline src/numeric/src/cumprod.cairo:19]: revert

File: src/numeric/src/cumsum.cairo [hunk 10]
```diff
@@ -10,18 +12,13 @@ fn cumsum<T, +Add<T>, +Copy<T>, +Drop<T>,>(mut sequence: Span<T>) -> Array<T> {
...
-                prev_value = sum;
-            },
-            Option::None => { break; },
-        };
+    let mut array = array![prev_value];
+    while !sequence.is_empty() {
+        let current_value = *sequence.pop_front().unwrap();
+        let sum = current_value + prev_value;
+        array.append(sum);
+        prev_value = sum;
```
Comment (LucasLvy) [inline src/numeric/src/cumsum.cairo:21]: revert

File: src/numeric/src/diff.cairo [hunk 11]
```diff
@@ -12,18 +12,13 @@ fn diff<T, +PartialOrd<T>, +Sub<T>, +Copy<T>, +Drop<T>, +Zeroable<T>,>(
...
-                prev_value = *current_value;
-            },
-            Option::None => { break; },
-        };
+    let mut array = array![Zeroable::zero()];
+    while !sequence.is_empty() {
+        let current_value = *sequence.pop_front().unwrap();
+        assert(current_value >= prev_value, 'Sequence must be sorted');
+        array.append(current_value - prev_value);
+        prev_value = current_value;
```
Comment (LucasLvy) [inline src/numeric/src/diff.cairo:21]: revert

**Review by LucasLvy — CHANGES_REQUESTED** [review]
can you also change all the `assert(x == y, 'xyz')` into `assert_eq!(x, y, "xyz");` you can use a regex for that

File: src/ascii/src/integer.cairo (hunk 2 above)
Comment (delaaxe) [reply src/ascii/src/integer.cairo]: done

File: src/bytes/src/bytes copy.cairo (hunk 3 above)
Comment (delaaxe) [reply src/bytes/src/bytes copy.cairo]: mistake sorry, removed

File: src/linalg/src/kron.cairo (hunk 5 above)
Comment (delaaxe) [reply src/linalg/src/kron.cairo]: done

File: src/linalg/src/norm.cairo (hunk 6 above)
Comment (delaaxe) [reply src/linalg/src/norm.cairo]: done

File: src/math/src/gcd_of_n_numbers.cairo (hunk 7 above)
Comment (delaaxe) [reply src/math/src/gcd_of_n_numbers.cairo]: done

File: src/math/src/lcm_of_n_numbers.cairo (hunk 8 above)
Comment (delaaxe) [reply src/math/src/lcm_of_n_numbers.cairo]: done

File: src/numeric/src/cumprod.cairo (hunk 9 above)
Comment (delaaxe) [reply src/numeric/src/cumprod.cairo]: done

File: src/numeric/src/cumsum.cairo (hunk 10 above)
Comment (delaaxe) [reply src/numeric/src/cumsum.cairo]: done

File: src/numeric/src/diff.cairo (hunk 11 above)
Comment (delaaxe) [reply src/numeric/src/diff.cairo]: done

File: Scarb.toml (hunk 1 above)
Comment (delaaxe) [reply Scarb.toml:22]: but i can udate it too if you prefer

**Comment by delaaxe** [comment]
> can you also change all the `assert(x == y, 'xyz')` into `assert_eq!(x, y, "xyz");` you can use a regex for that

I would wait for snfoundry to be ready before doing it in library code because error message catching infra isn't fully implemented yet https://github.com/foundry-rs/starknet-foundry/issues/1627

**Comment by delaaxe** [comment]
done as much as i could, it failed in many places (out of gas, formatter unimplemented, other weird errors)

**Comment by LucasLvy** [comment]
Let's not care about other tools, if someone wants to use it with foundry they'll use the old version

**Comment by delaaxe** [comment]
well we at argent would like to use the latest alexandria with foundry, but the assert! macros feel a bit too unstable to be shipped to mainnet

File: Scarb.toml (hunk 1 above)
Comment (delaaxe) [reply Scarb.toml]: done

**Review by LucasLvy — APPROVED** [review]
